### DISCUSSION TITLE
Test hangs when using mock.using (Automock)

### DISCUSSION THREAD

**Discussion by rienheuver** [automock/automock#265] [Q&A]
I have the following test that will not properly execute. It doesn't fail and it doesn't succeed, it just hangs forever. If I don't mock MikroORM it will continue (and crash because of the lacking mock), but MikroORM appears to be initialized just fine. Any clue on what's going on?

```
describe('MyClassGateway', () => {
  let myClassGateway: MyClassGateway;
  let myClassService: jest.Mocked<MyClassService>;

  beforeAll(async () => {
    const orm = await MikroORM.init({
      connect: false,
      type: 'mysql',
      dbName: 'test',
      entities: [MyClass],
    });

    const { unit, unitRef } = TestBed.create(MyClassGateway)
      .mock(MikroORM)
      .using(orm)
      .compile();

    myClassGateway = unit;
    myClassService = unitRef.get(MyClassService);
  });

  describe('createMyClass', () => {
    it('should return the created cause', () => {
      const mockMyInstance = {name: 'myClass-name', id: 0};
      myClassService.createMyClass.mockResolvedValue(mockMyInstance);
      const myInstance = myClassGateway.createMyClass(
        mockMyInstance
      );
      expect(myInstance).toEqual(mockMyInstance);
    });
  });
});
```

**Comment by omermorad** [comment]
Hello, @rienheuver

It's an interesting strategy that you're attempting to replace the mock with the actual MikroORM instance. Although this isn't quite what Automock intended for 😅, I suppose it has to do with the connection. Could you give some more context? Or perhaps a repository that helps to reproduce of the behavior?

**Reply by rienheuver** [reply]
I understand this is not really what automock is meant for, but it would be nice to use automock for our tests though. I instantiate MikroORM itself because the functions I am testing are decorated with `@useRequestContext()` from MikroORM. That decorator will throw an exception if there is no MikroORM-instance present. So the easiest mocking solution I thought of was to initialize MikroORM without letting it connect to a db.
Hopefully that gives it some more context!

**Comment by omermorad** [comment]
> I understand this is not really what automock is meant for, but it would be nice to use automock for our tests though. I instantiate MikroORM itself because the functions I am testing are decorated with `@useRequestContext()` from MikroORM. That decorator will throw an exception if there is no MikroORM-instance present. So the easiest mocking solution I thought of was to initialize MikroORM without letting it connect to a db. Hopefully that gives it some more context!

I understand. If you are experiencing this problem, I assume others are as well. Let's attempt to manage that scenario; could you kindly provide the code that uses `@useRequestContext()`? So that we can reproduce this, I would like to try running it on my machine.

**Reply by rienheuver** [reply]
Sure thing! I'm not sure if the following suffices though, so let me know if I need to provide more. For context: this is an api that is accessed over websockets (socket.io)

This is the function that has the decorator:
```ts
/**
 * Create a new cause
 * @param data
 * @param client
 */
@SubscribeMessage(BrainstormClientEvent.CreateCause)
@UseRequestContext()
public async createCause(
  @MessageBody() data: CreateCauseDto,
  @ConnectedSocket() client: Socket
): Promise<Cause> {
  if (
    !(await this.causesService.canCreate(
      client.participantId,
      client.sessionId,
      data.risk
    ))
  ) {
    throw new UnauthorizedException();
  }
  const cause = await this.causesService.createCause(
    data,
    client.participantId
  );
  client.broadcast
    .in(roomName(client.sessionId))
    .emit(BrainstormServerEvent.CauseCreated, cause);
  return cause;
}
```

**Comment by omermorad** [comment]
@rienheuver any update on this?

**Reply by rienheuver** [reply]
No, after my last update we didn't get any next questions or other responses. So because we couldn't use automock we've since decided to implement it differently. Don't know how anymore exactly since it's so long ago and part of the work a colleague did
